src/librbd/operation/MigrateRequest.cc

```diff
@@ -68,25 +101,129 @@ bool MigrateRequest<I>::should_complete(int r) {
   if (r < 0) {
     lderr(cct) << "encountered error: " << cpp_strerror(r) << dendl;
   }
+
   return true;
 }
 
 template <typename I>
-void MigrateRequest<I>::send_op() {
+void MigrateRequest<I>::copyup_objects() {
   I &image_ctx = this->m_image_ctx;
-  assert(image_ctx.owner_lock.is_locked());
   CephContext *cct = image_ctx.cct;
-  ldout(cct, 5) << dendl;
+  assert(image_ctx.owner_lock.is_locked());
+
+  if (image_ctx.size == 0) {
+    deep_copy_objects();
+    return;
+  }
 
+  ldout(cct, 5) << "from 0 to " << m_image_size_objects << dendl;
+
+  auto ctx = create_context_callback<
+    MigrateRequest<I>, &MigrateRequest<I>::handle_copyup_objects>(this);
+
+  assert(m_total_objects >= m_image_size_objects);
+  m_copyup_prog_ctx = new CopyupProgressContext(m_prog_ctx, m_total_objects);
   typename AsyncObjectThrottle<I>::ContextFactory context_factory(
     boost::lambda::bind(boost::lambda::new_ptr<C_MigrateObject<I> >(),
-      boost::lambda::_1, &image_ctx, m_snapc, boost::lambda::_2));
+      boost::lambda::_1, &image_ctx, image_ctx.snapc, boost::lambda::_2));
   AsyncObjectThrottle<I> *throttle = new AsyncObjectThrottle<I>(
-    this, image_ctx, context_factory, this->create_callback_context(), &m_prog_ctx,
-    0, m_overlap_objects);
+    this, image_ctx, context_factory, ctx, m_copyup_prog_ctx, 0,
+    m_image_size_objects);
   throttle->start_ops(image_ctx.concurrent_management_ops);
 }
 
+template <typename I>
+void MigrateRequest<I>::handle_copyup_objects(int r) {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+  ldout(cct, 5) << "r=" << r << dendl;
+
+  delete m_copyup_prog_ctx;
+  m_copyup_prog_ctx = nullptr;
+
+  if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed to copyup objects: " << cpp_strerror(r) << dendl;
+    this->complete(r);
+    return;
+  }
+
+  deep_copy_objects();
+}
+
+template <typename I>
+void MigrateRequest<I>::deep_copy_objects() {
+  I &image_ctx = this->m_image_ctx;
+  CephContext *cct = image_ctx.cct;
+
+  if (m_image_size_objects == m_total_objects) {
+    this->complete(0);
+    return;
+  }
+
+  assert(m_total_objects > m_image_size_objects);
+
+  ldout(cct, 5) << "from " << m_image_size_objects << " to " << m_total_objects
+                << dendl;
+
+  assert(image_ctx.parent != nullptr);
+
+  // If we are here snap_seqs should contain CEPH_NOSNAP and at least one
+  // snapshot, which size is larger than image size.
+  assert(m_snap_seqs.size() > 1);
+  auto snap_id_end = (++m_snap_seqs.rbegin())->first;
+
+  deep_copy::ObjectNumber object_number;
+  if (m_image_size_objects > 0) {
+    object_number = m_image_size_objects - 1;
+  }
+
+  auto ctx = create_context_callback<
+    MigrateRequest<I>, &MigrateRequest<I>::handle_deep_copy_objects>(this);
+
+  auto req = deep_copy::ImageCopyRequest<I>::create(
+      image_ctx.parent, &image_ctx, 0, snap_id_end, object_number, m_snap_seqs,
```

Comment:
So this just loops through the full object again?

Reply:
Do you mean "through the full _image_ again"? I am starting to deep copy the image from `object_number = m_image_size_objects - 1`.